### PR TITLE
feat: add init `dataKeys` & `dataValues`

### PR DESCRIPTION
**Pull request by b00ste**
# What does this PR introduce?

This PR allows to set `dataKeys` & `dataValues` during HypSP7 & HypLSP8 initialization.
Adds tests to check that the data keys are updated correctly.

### CODE REVIEW

File: src/HypLSP7.sol [hunk 1]
```diff
@@ -57,9 +58,9 @@ contract HypLSP7 is LSP7DigitalAssetInitAbstract, TokenRouter {
             isNonDivisible_: false // isNonDivisible set to `false` as not used anyway since decimals() is overriden
          });
 
-        // emit `DataChanged` event only if some metadata bytes is provided to save gas
-        if (_lsp4Metadata.length > 0) {
-            _setData(_LSP4_METADATA_KEY, _lsp4Metadata);
+        // set init data keys & values
+        if (dataKeys.length > 0 && dataKeys.length == dataValues.length) {
```
Comment (CJ42) [inline src/HypLSP7.sol:62]: Hmmm is this second check needed?
If I provide 2 x data keys and 3 x data values for instance, this check from `ERC725Y` will be skipped.

https://github.com/ERC725Alliance/ERC725/blob/5b9ff11678bcaba4ce31d4699b9f9fa484b3251d/implementations/contracts/ERC725Y.sol#L181-L194

image

Comment (CJ42) [reply src/HypLSP7.sol:62]: Suggest adding a test for this case above. That it should revert in case

File: test/HypLSP7.t.sol [hunk 2]
```diff
@@ -28,7 +28,23 @@ import { PausableCircuitBreakerHook } from "../src/ISM/PausableCircuitBreakerHoo
 import { IERC725Y } from "@erc725/smart-contracts/contracts/interfaces/IERC725Y.sol";
 
 // constants
-import { _LSP4_METADATA_KEY } from "@lukso/lsp4-contracts/contracts/LSP4Constants.sol";
+import { _INTERFACEID_LSP0 } from "@lukso/lsp0-contracts/contracts/LSP0Constants.sol";
```
Comment (CJ42) [inline test/HypLSP7.t.sol:31]: Ah ok, since `@lukso/lsp0-contracts` is just used for testing, suggest putting this under the `devDependencies` then (so it is not shipped in the package build).

File: test/HypLSP7.t.sol [hunk 3]
```diff
@@ -317,6 +335,20 @@ abstract contract HypTokenTest is Test {
 
         assertEq(aliceBalance, localToken.balanceOf(ALICE));
     }
+
+    function _getInitDataKeysAndValues() internal view returns (bytes32[] memory dataKeys, bytes[] memory dataValues) {
+        dataKeys = new bytes32[](4);
```
Comment (CJ42) [inline test/HypLSP7.t.sol:340]: Suggest adding comment:

```
// setting data keys for the following:
// - 1 x creator in the creator array 
// - creator's info under the map key
// - the token metadata
```

File: test/HypLSP7.t.sol [hunk 4]
```diff
@@ -346,11 +379,34 @@ contract HypLSP7Test is HypTokenTest {
...
+        hypLSP7Token.setData(_LSP4_TOKEN_NAME_KEY, name);
+    }
+
+    function test_SetData_ChangeTokenSymbol_Reverts(bytes memory name) public {
+        vm.prank(OWNER);
+        vm.expectRevert(LSP4TokenSymbolNotEditable.selector);
+        hypLSP7Token.setData(_LSP4_TOKEN_SYMBOL_KEY, name);
+    }
+
+    function test_SetData_ChangeTokenType_Reverts(bytes memory name) public {
```
Comment (CJ42) [inline test/HypLSP7.t.sol:398]: good tests 👌🏻

File: test/HypLSP7.t.sol
Comment (CJ42) [inline test/HypLSP7.t.sol:1]: I think we should also test that it is skipped and still initialize when you provide no data keys / values (2 empty arrays `[]`)

File: test/HypLSP7.t.sol (hunk 2 above)
Comment (b00ste) [reply test/HypLSP7.t.sol:31]: after installing `@lukso/lsp0-contracts` for testing purposes, I couldn't build, due to missing lsp17 package, that's why  added it. Will move both of them to devDeps

File: test/HypLSP7.t.sol
Comment (b00ste) [reply test/HypLSP7.t.sol:1]: good idea

**Review by CJ42 — APPROVED** [review]
